Userland/Libraries/LibJS/SourceCode.cpp

```diff
@@ -32,18 +33,50 @@ String const& SourceCode::code() const
     return m_code;
 }
 
+void SourceCode::compute_line_break_offsets() const
+{
+    if (m_code.is_empty())
+        return;
+
+    bool previous_code_point_was_carriage_return = false;
+    Utf8View view(m_code.view());
+    for (auto it = view.begin(); it != view.end(); ++it) {
+        u32 code_point = *it;
+        bool is_line_terminator = code_point == '\r' || (code_point == '\n' && !previous_code_point_was_carriage_return) || code_point == LINE_SEPARATOR || code_point == PARAGRAPH_SEPARATOR;
+        previous_code_point_was_carriage_return = code_point == '\r';
+
+        if (is_line_terminator)
+            m_line_break_offsets.append(view.byte_offset_of(it));
+    }
+}
+
 SourceRange SourceCode::range_from_offsets(u32 start_offset, u32 end_offset) const
 {
+    if (m_code.is_empty())
+        return { *this, {}, {} };
+
+    if (m_line_break_offsets.is_empty())
+        compute_line_break_offsets();
+
+    size_t line = 1;
+    size_t nearest_line_break_index = 0;
+    size_t nearest_preceding_line_break_offset = 0;
+
+    if (!m_line_break_offsets.is_empty()) {
+        binary_search(m_line_break_offsets, start_offset, &nearest_line_break_index);
+        line = 1 + nearest_line_break_index;
+        nearest_preceding_line_break_offset = m_line_break_offsets[nearest_line_break_index];
+    }
+
     Position start;
     Position end;
 
-    size_t line = 1;
     size_t column = 1;
 
     bool previous_code_point_was_carriage_return = false;
 
-    Utf8View view(m_code);
-    for (auto it = view.begin(); it != view.end(); ++it) {
+    Utf8View view(m_code.view());
+    for (auto it = view.iterator_at_byte_offset(nearest_preceding_line_break_offset); it != view.end(); ++it) {
```

Comment:
Not sure if it'll be an improvement, but I wonder if we could instead store the `Utf8View` on the class, then change `m_line_break_offsets` to be a `Vector<Utf8CodePointIterator>`. 

`iterator_at_byte_offset` makes me :grimacing: a bit because it's a `for` loop from the beginning of the string. Having a known iterator would avoid that, but it might also be totally inconsequential.

Reply:
Oh yeah, let's add a `iterator_at_byte_offset_without_validation` for situations like this where we know the byte offset is a valid code point boundary in the stream.